File: common/src/main/java/de/hype/bbsentials/shared/objects/BBServiceData.java
package de.hype.bbsentials.shared.objects;

import de.hype.bbsentials.server.objects.BBUser;
import de.hype.bbsentials.shared.constants.StatusConstants;
import de.hype.bbsentials.shared.constants.TradeType;
import org.jetbrains.annotations.NotNull;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

public class BBServiceData {
    protected int serviceId;
    protected StatusConstants status = StatusConstants.OPEN;
    protected String description;
    protected TradeType type;
    protected String title;
    protected BBUser hoster;
    protected List<Helper> helpers;
    protected List<Participiant> participants = new ArrayList<>();
    protected List<Participiant> queue = new ArrayList<>();
    protected int price;
    protected String dcMessageID;
    protected int maxUsers;
    protected boolean joinLock;
    protected boolean forceModOnline;
    public boolean circulateParticipants;

    public BBServiceData(String description, BBUser hoster, int price, List<Helper> helpers, int maxUsers, boolean forceModOnline) {
        this(null, description, hoster, price, helpers, maxUsers, forceModOnline);
    }

    public BBServiceData(@NotNull TradeType type, BBUser hoster, int price, List<Helper> helpers, boolean forceModOnline) {
        this(type, type.getDescription(), hoster, price, helpers, type.getMaximumUsers(helpers.size()), forceModOnline);
    }

    public BBServiceData(TradeType type, String description, BBUser hoster, int price, List<Helper> helpers, int maxUsers, boolean forceModOnline) {
        this.type = type;
        this.hoster = hoster;
        this.helpers = helpers;
        this.price = price;
        this.maxUsers = maxUsers;
        this.forceModOnline = forceModOnline;
        this.description = description;
    }

    public int getServiceId() {
        return serviceId;
    }

    public static class Participiant {
        public BBUser user;
        public boolean priority;
        public boolean autoRequeue;
        public Instant joinTime;
        public int price;

        public Participiant(BBUser user, boolean priority, boolean free, BBServiceData data) {
            this(user, priority, free, Instant.now(), false, data);
        }

        public Participiant(BBUser user, boolean priority, boolean free, Instant joinTime, boolean autoRequeue, BBServiceData data) {
            this(user, priority, joinTime, autoRequeue, free ? 0 : data.price);
        }

        public Participiant(BBUser user, boolean priority, Instant joinTime, boolean autoRequeue, int price) {
            this.user = user;
            this.priority = priority;
            this.price = price;
            this.autoRequeue = autoRequeue;
            this.joinTime = joinTime;
        }

        @Override
        public boolean equals(Object obj) {
            if (obj instanceof Participiant participiant) return participiant.user.equals(user);
            if (obj instanceof BBUser bbUser) return bbUser.equals(this.user);
            return false;
        }

        @Override
        public int hashCode() {
            return user.hashCode();
        }

        public String getDiscordParticipiantString() {
            String string = user.getMcUsername();
            if (price == 0) string = "**%s**".formatted(string);
            if (priority) string = "__%s__".formatted(string);
            if (autoRequeue) string = "*%s*".formatted(string);
            return string;

        }
    }

    public static class Helper {
        public BBUser user;
        public String username;

        public Helper(BBUser user) {
            this.user = user;
            this.username = user.getMcUsername();
        }

        public Helper(BBUser user, String username) {
            this.user = user;
            this.username = username;
        }

        @Override
        public boolean equals(Object obj) {
            if (obj instanceof Helper helper) return helper.username.equals(username);
            if (obj instanceof String string) {
                if (username != null) return string.equalsIgnoreCase(username);
                else return user.getMcUsername().equalsIgnoreCase(string);
            }
            if (obj instanceof BBUser bbUser) {
                if (user != null) return bbUser.equals(user);
                else return bbUser.getMcUsername().equalsIgnoreCase(username);
            }
            return false;
        }

        @Override
        public int hashCode() {
            if (username != null) return username.hashCode();
            else return user.getMcUsername().hashCode();
        }
    }
}

File: common/src/main/kotlin/de/hype/bingonet/shared/objects/BBServiceData.kt
package de.hype.bingonet.shared.objects

import de.hype.bingonet.server.objects.BBUser
import de.hype.bingonet.shared.constants.StatusConstants
import de.hype.bingonet.shared.constants.TradeType
import java.time.Instant

open class BBServiceData(
    @JvmField protected val type: TradeType?,
    open var description: String,
    open var hoster: BBUser,
    @JvmField var price: Int,
    @JvmField var helpers: MutableList<Helper>,
    @JvmField var maxUsers: Int,
    @JvmField var forceModOnline: Boolean
) {
    var serviceId: Int = 0

    @JvmField
    var status: StatusConstants = StatusConstants.OPEN

    @JvmField
    var title: String? = null

    open var participants: MutableList<Participant> = ArrayList<Participant>()

    @JvmField
    var queue: MutableList<Participant> = ArrayList<Participant>()
    protected lateinit var dcMessageID: String

    @JvmField
    var joinLock: Boolean = false

    @JvmField
    var circulateParticipants: Boolean = false

    constructor(
        type: TradeType,
        hoster: BBUser,
        price: Int,
        helpers: MutableList<Helper>,
        forceModOnline: Boolean
    ) : this(type, type.description, hoster, price, helpers, type.getMaximumUsers(helpers.size), forceModOnline)

    class Participant(
        @JvmField var user: BBUser,
        @JvmField var priority: Boolean,
        @JvmField var joinTime: Instant,
        @JvmField var autoRequeue: Boolean,
        @JvmField var price: Int
    ) {
        constructor(user: BBUser, priority: Boolean, free: Boolean, data: BBServiceData) : this(
            user,
            priority,
            free,
            Instant.now(),
            false,
            data
        )

        constructor(
            user: BBUser,
            priority: Boolean,
            free: Boolean,
            joinTime: Instant,
            autoRequeue: Boolean,
            data: BBServiceData
        ) : this(user, priority, joinTime, autoRequeue, if (free) 0 else data.price)

        override fun equals(other: Any?): Boolean {
            if (other is Participant) return other.user == user
            if (other is BBUser) return other == this.user
            return false
        }

        override fun hashCode(): Int {
            return user.hashCode()
        }

        val discordParticipantString: String
            get() {
                var string = user.getMcUsername()
                if (price == 0) string = "**$string**"
                if (priority) string = "__${string}__"
                if (autoRequeue) string = "*$string*"
                return string
            }
    }

    class Helper {
        private var user: BBUser?
        private var username: String?

        constructor(user: BBUser) {
            this.user = user
            this.username = user.getMcUsername()
        }

        constructor(user: BBUser?, username: String) {
            this.user = user
            this.username = username
        }

        override fun equals(other: Any?): Boolean {
            if (other is Helper) return other.username == username
            if (other is String) {
                if (username != null) return other.equals(username, ignoreCase = true)
                else return user!!.getMcUsername().equals(other, ignoreCase = true)
            }
            if (other is BBUser) {
                if (user != null) return other == user
                else return other.getMcUsername().equals(username, ignoreCase = true)
            }
            return false
        }

        override fun hashCode(): Int {
            if (username != null) return username.hashCode()
            else return user!!.getMcUsername().hashCode()
        }

        fun getUserName(): String {
            return username ?: user!!.mcUsername
        }

        fun getUser(): BBUser? {
            return user
        }
    }
}
